sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala

```diff
@@ -184,19 +175,23 @@ class DataSourceV2Strategy(session: SparkSession) extends Strategy with Predicat
         newSchema, catalog.asTableCatalog, ident, "CREATE TABLE")
 
       CreateTableExec(catalog.asTableCatalog, ident, structTypeToV2Columns(newSchema),
-        partitioning, qualifyLocInTableSpec(tableSpec), ifNotExists) :: Nil
+        partitioning, qualifyLocInTableSpec(session, tableSpec), ifNotExists) :: Nil
 
     case CreateTableAsSelect(ResolvedIdentifier(catalog, ident), parts, query, tableSpec,
-        options, ifNotExists, analyzedQuery) =>
+        options, ifNotExists, analyzedQuery, table, write) =>
       assert(analyzedQuery.isDefined)
       val writeOptions = new CaseInsensitiveStringMap(options.asJava)
       catalog match {
         case staging: StagingTableCatalog =>
-          AtomicCreateTableAsSelectExec(staging, ident, parts, analyzedQuery.get, planLater(query),
-            qualifyLocInTableSpec(tableSpec), writeOptions, ifNotExists) :: Nil
+          assert(table.isDefined)
+          assert(write.isDefined)
+          AtomicCreateTableAsSelectExec(
```

Comment:
This PR focuses on atomic CTAS/RTAS as we can safely stage a new table in the optimizer and then write to it during the execution. I don't have a good solution for non-atomic CTAS/RTAS. Any ideas are appreciated.

Reply:
we can probably follow v1 write and trigger a nested execution to do table insertion in CTAS/RTAS command: https://github.com/apache/spark/commit/4b40920e33176fc8b18380703e4dcf4d16824094

Reply:
It seems we can do the same thing for atomic CTAS/RTAS as well: create the staging table, build a write plan with it, and call `session.sessionState.executePlan`

Reply:
Let me explore this.

Reply:
Resolving as the logic has been changed.